### PR TITLE
[CLI] Add `aptos move show` subcommand group

### PR DESCRIPTION
**Pull request by banool**
### Description
Occasionally we'll want to look at the ABI that comes directly out of the compiler. Previously when we wanted to do this we'd hack up a script for that one instance and then throw it away. This PR adds support for this in the CLI with the `aptos move show` subcommand group, starting with `aptos move show abi`.

I have marked this subcommand group has hidden for now, as the difference between this ABI and the ABI returned by the API might confuse some people.

### Test Plan
```
$ cargo run -p aptos -- move show abi --help
aptos-move-show-abi 1.0.8
Compile a package and show information about the ABIs of the compiled modules.

For example, this would show the function `transfer` in the module `coin`:

aptos move show abi --modules coin --names transfer

USAGE:
    aptos move show abi [OPTIONS]

OPTIONS:
        --bytecode-version <BYTECODE_VERSION>
            Specify the version of the bytecode the compiler is going to emit

    -h, --help
            Print help information

        --included-artifacts <INCLUDED_ARTIFACTS>
            Artifacts to be generated when building the package

            Which artifacts to include in the package. This can be one of `none`, `sparse`, and
            `all`. `none` is the most compact form and does not allow to reconstruct a source
            package from chain; `sparse` is the minimal set of artifacts needed to reconstruct a
            source package; `all` includes all available artifacts. The choice of included artifacts
            heavily influences the size and therefore gas cost of publishing: `none` is the size of
            bytecode alone; `sparse` is roughly 2 times as much; and `all` 3-4 as much.

            [default: sparse]

        --modules <MODULES>...
            If provided, only show items from the given Move modules. These should be module names,
            not file paths. For example, `coin`

        --named-addresses <NAMED_ADDRESSES>
            Named addresses for the move binary

            Example: alice=0x1234, bob=0x5678

            Note: This will fail if there are duplicates in the Move.toml file remove those first.

            [default: ]

        --names <NAMES>...
            If provided, only show items with the given names. For example, `transfer`

        --output-dir <OUTPUT_DIR>
            Path to save the compiled move package

            Defaults to `<package_dir>/build`

        --package-dir <PACKAGE_DIR>
            Path to a move package (the folder with a Move.toml file)

        --skip-fetch-latest-git-deps
            Skip pulling the latest git dependencies

            If you don't have a network connection, the compiler may fail due to no ability to pull
            git dependencies.  This will allow overriding this for local development.

    -V, --version
            Print version information
```
```
cargo run -p aptos -- move show abi --package-dir /Users/dport/a/core/aptos-move/framework/aptos-framework --modules coin
```
```
{
  "Result": [
    {
      "EntryFunction": {
        "name": "transfer",
        "module_name": {
          "address": "0000000000000000000000000000000000000000000000000000000000000001",
          "name": "coin"
        },
        "doc": " Transfers `amount` of coins `CoinType` from `from` to `to`.",
        "ty_args": [
          {
            "name": "coin_type"
          }
        ],
        "args": [
          {
            "name": "to",
            "type_tag": "address"
          },
          {
            "name": "amount",
            "type_tag": "u64"
          }
        ]
      }
    },
    {
      "EntryFunction": {
        "name": "upgrade_supply",
        "module_name": {
          "address": "0000000000000000000000000000000000000000000000000000000000000001",
          "name": "coin"
        },
        "doc": " Upgrade total supply to use a parallelizable implementation if it is\n available.",
        "ty_args": [
          {
            "name": "coin_type"
          }
        ],
        "args": []
      }
    }
  ]
}
```

```
cargo run -p aptos -- move show abi --package-dir /Users/dport/a/core/aptos-move/framework/aptos-framework --modules vesting --names update_operator unlock_rewards
```
```
{
  "Result": [
    {
      "EntryFunction": {
        "name": "unlock_rewards",
        "module_name": {
          "address": "0000000000000000000000000000000000000000000000000000000000000001",
          "name": "vesting"
        },
        "doc": " Unlock any accumulated rewards.",
        "ty_args": [],
        "args": [
          {
            "name": "contract_address",
            "type_tag": "address"
          }
        ]
      }
    },
    {
      "EntryFunction": {
        "name": "update_operator",
        "module_name": {
          "address": "0000000000000000000000000000000000000000000000000000000000000001",
          "name": "vesting"
        },
        "doc": "",
        "ty_args": [],
        "args": [
          {
            "name": "contract_address",
            "type_tag": "address"
          },
          {
            "name": "new_operator",
            "type_tag": "address"
          },
          {
            "name": "commission_percentage",
            "type_tag": "u64"
          }
        ]
      }
    }
  ]
}
```

### CODE REVIEW

File: crates/aptos/src/move_tool/mod.rs [hunk 1]
```diff
@@ -77,48 +78,51 @@ use transactional_tests_runner::TransactionalTestOpts;
...
+    Init(InitPackage),
     List(ListPackage),
-    Clean(CleanPackage),
-    VerifyPackage(VerifyPackage),
+    Prove(ProvePackage),
+    Publish(PublishPackage),
     Run(RunFunction),
     RunScript(RunScript),
+    #[clap(subcommand, hide = true)]
+    Show(show::ShowTool),
```
Comment (gregnazario) [inline crates/aptos/src/move_tool/mod.rs:96]: Please put refactors in separate commits

Comment (gregnazario) [inline crates/aptos/src/move_tool/mod.rs:96]: Why is it a hide?

File: crates/aptos/src/move_tool/show.rs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+pub struct Abi {
+    /// If provided, only show items from the given Move modules. These should be module
+    /// names, not file paths. For example, `coin`.
+    #[clap(long, multiple_values = true)]
+    modules: Vec<String>,
+
+    /// If provided, only show items with the given names. For example, `transfer`.
+    #[clap(long, multiple_values = true)]
+    names: Vec<String>,
+
```
Comment (gregnazario) [inline crates/aptos/src/move_tool/show.rs:43]: How do you envision this if the ABI format changes?

Mostly because ABI here is only for entry functions, which can be confusing if something is done around view functions and others.

File: crates/aptos/src/move_tool/show.rs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+use anyhow::Context;
+use aptos_framework::{BuildOptions, BuiltPackage};
+use aptos_types::transaction::EntryABI;
+use async_trait::async_trait;
+use clap::{Parser, Subcommand};
+
+/// Commands for showing information about compilation artifacts.
+#[derive(Subcommand)]
+pub enum ShowTool {
+    /// Compile the package and show information about the ABIs of the compiled modules.
```
Comment (gregnazario) [inline crates/aptos/src/move_tool/show.rs:15]: Don't need this comment here, the Abi tool will be used for the CLI help

File: crates/aptos/src/move_tool/show.rs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    included_artifacts_args: IncludedArtifactsArgs,
+
+    #[clap(flatten)]
+    move_options: MovePackageDir,
+}
+
+#[async_trait]
+impl CliCommand<Vec<EntryABI>> for Abi {
+    fn command_name(&self) -> &'static str {
+        "Abi"
```
Comment (gregnazario) [inline crates/aptos/src/move_tool/show.rs:54]: Similiarly `ShowAbi` to prevent confusion with Abis in general

File: crates/aptos/src/move_tool/mod.rs (hunk 1 above)
Comment (banool) [reply crates/aptos/src/move_tool/mod.rs:96]: Will do.

Comment (banool) [reply crates/aptos/src/move_tool/mod.rs:96]: See summary.

File: crates/aptos/src/move_tool/show.rs (hunk 3 above)
Comment (banool) [reply crates/aptos/src/move_tool/show.rs:15]: Neato.

File: crates/aptos/src/move_tool/show.rs (hunk 2 above)
Comment (banool) [reply crates/aptos/src/move_tool/show.rs:43]: Which part of this do you mean, the filtering? If the ABI changes it'll require wholesale changes in a lot of places, not sure this is special in any way?

**Comment by wrwg** [comment]
I'm a bit confused here. Aren't the compiler abis already stored in the build folder? Is this just displaying them?

Also pl include someone from the Move team in PRs related to the `aptos move` command group.

**Comment by banool** [comment]
Yeah this is just displaying them. Will tag the Move team in the future.

**Comment by banool** [comment]
All obvious feedback addressed.

File: crates/aptos/src/move_tool/show.rs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+                        }
+                    },
+                    EntryABI::TransactionScript(_) => {
+                        // If there were any modules specified we ignore scripts.
+                        if !self.modules.is_empty() {
+                            return false;
+                        }
+                    },
+                }
+                true
```
Comment (davidiw) [inline crates/aptos/src/move_tool/show.rs:103]:
```suggestion
                match &abi {
                    EntryABI::EntryFunction(func) => {
                        !(self.modules.is_empty()
                            || self
                                .modules
                                .contains(&func.module_name().name().to_string()))
                    },
                    EntryABI::TransactionScript(_) => {
                        // If there were any modules specified we ignore scripts.
                        !self.modules.is_empty()
                    },
                    _ => true
                }
```

the code can be slimmed down a bit

**Review by davidiw — APPROVED** [review]

File: crates/aptos/src/move_tool/show.rs (hunk 5 above)
Comment (banool) [reply crates/aptos/src/move_tool/show.rs:103]: Doesn't vibe with my brain as well, I like the series of `check -> return false` parts followed by a final `true`. But happy to make the change.
